From 15e4e42b8573fcd2d94b6c79a129ace3cc19e550 Mon Sep 17 00:00:00 2001
From: Jammy Huang <jammy_huang@aspeedtech.com>
Date: Thu, 18 Sep 2025 17:09:57 +0800
Subject: [PATCH 31/32] misc: sli_ast2700: Enhance pad delay calibration
 functions

Calibration could failed in case:
1. There are more than 1 window available.
2. The size of windows other than the last one are too small.
3. The last window exceeds the COARSE_END.

Modify the condition to pick last window that exceeds the COARSE_END.

Signed-off-by: Jammy Huang <jammy_huang@aspeedtech.com>
Change-Id: Ieef6edd817c6f537b6a1431ffb0297a93cd00a2b
(cherry picked from commit f6235a3644e47899804b0a20f7499d2cee37cd72)
Signed-off-by: Vasanthakumar K <vasanthakumark@ami.com>
---
 drivers/misc/sli_ast2700.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/misc/sli_ast2700.c b/drivers/misc/sli_ast2700.c
index 48bdae888d..d0d7b5fb0d 100644
--- a/drivers/misc/sli_ast2700.c
+++ b/drivers/misc/sli_ast2700.c
@@ -247,7 +247,7 @@ static void sli_calibrate_ahb_delay(struct sli_data *data)
 		}
 	}
 
-	if (win_size == 0 && d_last_pass != -1) {
+	if (d_last_pass != -1 && (d_last_pass - d_first_pass) > win_size) {
 		win_size = d_last_pass - d_first_pass;
 		sli_log_ahb_pad_delay(data, d_first_pass, d_last_pass);
 		debug("IOD SLIH DS coarse win: {%d, %d}\n", d_first_pass, d_last_pass);
@@ -440,7 +440,7 @@ static void sli_calibrate_mbus_delay(struct sli_data *data, bool is_k_rx)
 			}
 		}
 
-		if (win_size == 0 && d_last_pass != -1) {
+		if (d_last_pass != -1 && (d_last_pass - d_first_pass) > win_size) {
 			win_size = d_last_pass - d_first_pass;
 			sli_log_mbus_pad_delay(data, 0, d_first_pass, d_last_pass);
 			debug("%s SLIM DS coarse win: {%d, %d}\n", die_name, d_first_pass, d_last_pass);
@@ -584,7 +584,7 @@ static void sli_calibrate_video_delay(struct sli_data *data, bool is_DS, bool is
 		}
 	}
 
-	if (win_size == 0 && d_last_pass != -1) {
+	if (d_last_pass != -1 && (d_last_pass - d_first_pass) > win_size) {
 		win_size = d_last_pass - d_first_pass;
 		sli_log_video_pad_delay(scu, d_first_pass, d_last_pass);
 		debug("%s SLIV %s coarse win: {%d, %d}\n", die_name, dir_name, d_first_pass, d_last_pass);
-- 
2.34.1

